der/src/tag.rs

```diff
@@ -271,85 +287,159 @@ impl Tag {
+
+                Tag::ContextSpecific {
+                    constructed,
+                    number,
+                }
+            }
+            0xC0..=0xFF => {
+                let (constructed, number) = parse_parts(first_byte, reader)?;
+
+                Tag::Private {
+                    constructed,
+                    number,
+                }
+            }
+            // universal tag in long form
+            0x1F => return Err(ErrorKind::TagNumberInvalid.into()),
+            byte => return Err(ErrorKind::TagUnknown { byte }.into()),
+        };
+
+        Ok(tag)
     }
 }
 
-impl From<Tag> for u8 {
-    fn from(tag: Tag) -> u8 {
-        tag.octet()
+fn ensure_multi_byte(number: u32) -> Result<TagNumber> {
+    if number >= u32::from(TagNumber::MASK) {
+        Ok(TagNumber::new(number))
+    } else {
+        // must be in short form
+        Err(ErrorKind::TagNumberInvalid.into())
     }
 }
 
-impl From<&Tag> for u8 {
-    fn from(tag: &Tag) -> u8 {
-        u8::from(*tag)
+fn parse_parts<'a, R: Reader<'a>>(first_byte: u8, reader: &mut R) -> Result<(bool, TagNumber)> {
+    let constructed = first_byte & CONSTRUCTED_FLAG != 0;
+    let first_number_part = first_byte & TagNumber::MASK;
+
+    if first_number_part != TagNumber::MASK {
+        return Ok((constructed, TagNumber::new(first_number_part.into())));
     }
-}
 
-impl<'a> Decode<'a> for Tag {
-    type Error = Error;
+    let mut multi_byte_tag_number = 0;
 
-    fn decode<R: Reader<'a>>(reader: &mut R) -> Result<Self> {
-        reader.read_byte().and_then(Self::try_from)
+    for _ in 0..Tag::MAX_SIZE - 2 {
+        multi_byte_tag_number <<= 7;
+
+        let byte = reader.read_byte()?;
+        multi_byte_tag_number |= u32::from(byte & 0x7F);
+
+        if byte & 0x80 == 0 {
+            return Ok((constructed, ensure_multi_byte(multi_byte_tag_number)?));
+        } else if byte == 0x80 {
+            // leading zeros in long form
+            return Err(ErrorKind::TagNumberInvalid.into());
+        }
+    }
+
+    let byte = reader.read_byte()?;
+    if multi_byte_tag_number > u32::MAX >> 7 || byte & 0x80 != 0 {
+        return Err(ErrorKind::TagNumberInvalid.into());
     }
+    multi_byte_tag_number |= u32::from(byte & 0x7F);
+
+    Ok((constructed, ensure_multi_byte(multi_byte_tag_number)?))
 }
 
 impl Encode for Tag {
+    #[allow(clippy::cast_possible_truncation)]
     fn encoded_len(&self) -> Result<Length> {
-        Ok(Length::ONE)
+        let number = self.number().value();
+
+        let length = if number <= 30 {
+            Length::ONE
+        } else {
+            Length::new(number.ilog2() as u16 / 7 + 2)
+        };
+
+        Ok(length)
     }
 
     fn encode(&self, writer: &mut impl Writer) -> Result<()> {
-        writer.write_byte(self.into())
+        let mut first_byte = self.class() as u8 | u8::from(self.is_constructed()) << 5;
+
+        let number = self.number().value();
+
+        if number <= 30 {
```

Comment:
```suggestion
        if number < u32::from(TagNumber::MASK) {
```

Though I see `u32::from(TagNumber::MASK)` in this file enough that maybe it's worth a separate constant, e.g. `TagNumber::MAX_SINGLE_BYTE_TAG` or something? I don't feel super strongly, just a thought.

Reply:
I think one constant may be enough here, though that may be changed in the future is needed. Also used the constant in some other places.